compiler/rustc_parse/src/parser/mod.rs

```diff
@@ -839,6 +868,48 @@ impl<'a> Parser<'a> {
         Ok((v, trailing, recovered))
     }
 
+    fn recover_missing_braces_around_closure_body(
+        &mut self,
+        right_pipe_span: Span,
+        mut expect_err: DiagnosticBuilder<'_>,
+    ) -> PResult<'a, ()> {
+        let initial_semicolon = self.token.span;
+
+        expect_err.span_help(
+            initial_semicolon,
+            "This `;` turns the expression into a statement, which must be placed in a block",
```

Comment:
Somehow reading this message in the test output below doesn't get the point across for me. I'm not entirely sure why. I think it might be that "This `;`" is clear, but then I think, "what expression? what statement?". The problem might be that those concepts seem too abstract and detached from the actual code to be understood at a glance. I'm not exactly sure how to fix it though.

Reply:
I understand your pain. It took me a lot of time to write this sentence and I'm still not satisfied with it.

To mitigate your "what expression? what statement?" problem, I see two solutions:
  - reword the message as "This `;` turns the **preceding** expression into a statement [...]" (see first screenshot),
  - span the help message on the whole "expression+semicolon" chunk, changing the message to "the trailing `;` turns the expression into a statement" (see second screenshot).

Both are not ideal, though.

Overall, I'm not even sure if it is a good idea to expect from the user to know the difference between an expression and a statement, given that users may not have knowledge in language grammar.

Reply:
An option we have is to abuse the `MultiSpan` API:

```rust
let mut sp: MultiSpan = vec![initial_semicolon].into();
sp.push_label(initial_semicolon, "this `;` turns the preceding expression into a statement");
sp.push_label(expr.span, "this expression is a statement because of the trailing semicolon");
sp.push_label(closure.span, "this is the parsed closure...");
sp.push_label(end_brace_span, "...but likely you meant the closure to end here");
```

Reply:
While this definitely works, I think this adds a lot of noise in the `note` section:
![A screenshot of the error message with all the spanned labels suggested by estebank. The result is very cumbersome. We don't know what to read first.](https://user-images.githubusercontent.com/25402018/131692915-544911b5-5f22-48a6-ba85-4890cb36b53c.png)

I think the best solution here is to remove the `spanned_label` whose message is "this expression is a statement because of the trailing semicolon", but this needs more experiments.

Reply:
Another alternative is to add a new note for either the statement or the closure.

It could even make sense to put the "this is the parsed closure..."/"...but likely you meant the closure to end here" in its own note, which would make the output more vertically prominent, but easier to read.

Reply:
> It could even make sense to put the "this is the parsed closure..."/"...but likely you meant the closure to end here" in its own note, which would make the output more vertically prominent, but easier to read.

I implemented it and IMO this is what gives the best results! It makes the error message longer, but it is way easier to read, as each subdiagnostic relate a specific concept.